src/platform/storage_winkernel.c

```diff
@@ -506,3 +580,178 @@ CxPlatStorageReadValue(
+        return Status;
+    }
+
+    Status =
+        ZwDeleteValueKey(
+            Storage->RegKey,
+            NameString);
+
+    CXPLAT_FREE(NameString, QUIC_POOL_PLATFORM_TMP_ALLOC);
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatStorageClear(
+    _In_ CXPLAT_STORAGE* Storage
+    )
+{
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    ULONG InfoLength = 0;
+    ULONG AllocatedLength = 0;
+    PKEY_VALUE_BASIC_INFORMATION Info = NULL;
+
+    //
+    // Iterate through all values and delete them
+    // We always use index 0 because deletion shifts the remaining values
+    //
+    while (TRUE) {
+        //
+        // First, query the required buffer size for the value name
+        //
+        Status = ZwEnumerateValueKey(
+            Storage->RegKey,
+            0, // Always use index 0 since we delete as we go
+            KeyValueBasicInformation,
+            NULL,
+            0,
+            &InfoLength);
+
+        if (Status == STATUS_NO_MORE_ENTRIES) {
+            Status = QUIC_STATUS_SUCCESS;
+            break;
+        } else if (Status != STATUS_BUFFER_OVERFLOW && Status != STATUS_BUFFER_TOO_SMALL) {
+            if (QUIC_FAILED(Status)) {
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Status,
+                    "ZwEnumerateValueKey (size query) failed");
+                goto Exit;
+            }
+        }
+
+        //
+        // Allocate or reallocate buffer only if current buffer is too small
+        //
+        if (Info == NULL || InfoLength > AllocatedLength) {
+            if (Info != NULL) {
+                CXPLAT_FREE(Info, QUIC_POOL_PLATFORM_TMP_ALLOC);
+            }
+
+            Info = CXPLAT_ALLOC_PAGED(InfoLength, QUIC_POOL_PLATFORM_TMP_ALLOC);
+            if (Info == NULL) {
+                Status = QUIC_STATUS_OUT_OF_MEMORY;
+                QuicTraceEvent(
+                    AllocFailure,
+                    "Allocation of '%s' failed. (%llu bytes)",
+                    "KEY_VALUE_BASIC_INFORMATION",
+                    InfoLength);
+                goto Exit;
+            }
+            AllocatedLength = InfoLength;
+        }
+
+        //
+        // Get the value name
+        //
+        Status = ZwEnumerateValueKey(
+            Storage->RegKey,
+            0, // Always use index 0 since we delete as we go
+            KeyValueBasicInformation,
+            Info,
+            AllocatedLength,
+            &InfoLength);
+
+        if (Status == STATUS_NO_MORE_ENTRIES) {
+            Status = QUIC_STATUS_SUCCESS;
+            break;
+        } else if (QUIC_FAILED(Status)) {
+            QuicTraceEvent(
+                LibraryErrorStatus,
+                "[ lib] ERROR, %u, %s.",
+                Status,
+                "ZwEnumerateValueKey failed");
+            goto Exit;
+        }
+
+        CXPLAT_DBG_ASSERT(InfoLength == AllocatedLength);
```

Comment:
Is that true? If we allocate a big buffer then have some smaller data on the next iteration, I don't think it holds.